### PR TITLE
logs: make log level configurable through meshconfig

### LINKED ISSUE #3646: OSM Controller's log level should be configurable via MeshConfig
OpenServiceMesh.controllerLogLevel should be configurable via MeshConfig.
This is currently only configurable via Helm chart value and cannot be easily changed without reinstalling OSM.
Being able to change the log level of the **OSM Controller** will be very useful for debugging issues.

### PR DESCRIPTION
**Pull request by eduser25**
New `osmLogLevel` option has been added under observability meshconfig
spec (seems the most related and relevant to the available categories)
to allow changing the log level of OSM at runtime.

The initial log level flag has been kept as it still mandates boot log
level up to the point configurator and the log level handler are both
up, which then meshconfig value takes over.

The handler can't be easilly moved to the `logger` package as it
introduces a hard dependency cycle between logger<->configurator, hence
the handler has been left on osm-controller main.

Adds a unit test, and see below a working example on a running deployment.

Fixes #3646

Signed-off-by: Eduard Serra <eduser25@gmail.com>

**Description**:

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| Control Plane              | [x] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project?
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change?

### CODE REVIEW

**Comment by eduser25** [comment]
```
...
{"level":"info","component":"envoy/rds","time":"2021-06-23T19:07:17Z","file":"response.go:105","message":"RDS did not fulfill all requested resources (diff: Set{}). Fulfill with empty RouteConfigs."}
{"level":"info","component":"envoy/rds","time":"2021-06-23T19:07:17Z","file":"response.go:105","message":"RDS did not fulfill all requested resources (diff: Set{}). Fulfill with empty RouteConfigs."}
<only info logs til around now>
```

change to trace through kubectl edit meshconfig
```
{"level":"info","component":"osm-controller/main","time":"2021-06-23T19:07:58Z","file":"osm-controller.go:371","message":"Global log level changed to: trace"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:02Z","file":"rotor.go:51","message":"Cert ads will not be rotated; expires in 87599h58m50.3121094s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:02Z","file":"rotor.go:51","message":"Cert osm-validator.osm-system.svc will not be rotated; expires in 87599h58m50.429469332s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:02Z","file":"rotor.go:51","message":"Cert server-0-1.server-0.cluster.local will not be rotated; expires in 23h59m10.113151758s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:02Z","file":"rotor.go:51","message":"Cert server-0-0.server-0.cluster.local will not be rotated; expires in 23h59m10.272788317s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:07Z","file":"rotor.go:51","message":"Cert server-0-1.server-0.cluster.local will not be rotated; expires in 23h59m5.112912093s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:07Z","file":"rotor.go:51","message":"Cert server-0-0.server-0.cluster.local will not be rotated; expires in 23h59m5.272500951s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:07Z","file":"rotor.go:51","message":"Cert ads will not be rotated; expires in 87599h58m45.311669134s; renewBeforeCertExpires is 30s"}
{"level":"trace","component":"certificate/CertRotor","time":"2021-06-23T19:08:07Z","file":"rotor.go:51","message":"Cert osm-validator.osm-system.svc will not be rotated; expires in 87599h58m45.429139366s; renewBeforeCertExpires is 30s"}
....
```

reset to info through kubectl edit meshconfig
```
{"level":"info","component":"osm-controller/main","time":"2021-06-23T19:08:12Z","file":"osm-controller.go:371","message":"Global log level changed to: info"}
<no more trace logs>
```

File: charts/osm/crds/meshconfig.yaml [hunk 1]
```diff
@@ -155,6 +155,10 @@ spec:
                   description: Configuration for observing the service mesh, including metrics, logs, tracing etc,.
                   type: object
                   properties:
+                    osmLogLevel:
+                      description: Allows setting OSM's log level at runtime
```
Comment (shashankram) [inline charts/osm/crds/meshconfig.yaml:159]:
```suggestion
                      description: Allows setting OSM control plane log level at runtime
```

File: cmd/osm-controller/osm-controller.go [hunk 2]
```diff
@@ -79,7 +80,7 @@ var (
 )
 
 func init() {
-	flags.StringVarP(&verbosity, "verbosity", "v", "info", "Set log verbosity level")
+	flags.StringVarP(&verbosity, "verbosity", "v", constants.DefaultOSMLogLevel, "Set boot log verbosity level")
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:83]: If the initial logging is set using MeshConfig, why is this needed?

File: cmd/osm-controller/osm-controller.go [hunk 3]
```diff
@@ -333,3 +337,44 @@ func getOSMControllerPod(kubeClient kubernetes.Interface) (*corev1.Pod, error) {
 
 	return pod, nil
 }
+
+// StartGlobalLogLevelHandler registers a listener to meshconfig events and log level changes,
+// and applies new log level at global scope
+func StartGlobalLogLevelHandler(cfg configurator.Configurator, stop <-chan struct{}) {
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:343]: Move this to its own file `logging.go` for log management.

File: cmd/osm-controller/osm-controller.go [hunk 4]
```diff
@@ -333,3 +337,44 @@ func getOSMControllerPod(kubeClient kubernetes.Interface) (*corev1.Pod, error) {
...
+
+	go func() {
+		for {
+			select {
+			case <-meshConfigChannel:
+				logLevel := cfg.GetOSMLogLevel()
+				if logLevel != currentLogLevel {
+					err := logger.SetLogLevel(logLevel)
+					if err != nil {
+						log.Error().Msgf("Failed to set log level from meshconfig: %v", err)
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:369]: I tend to use `Error` in log messages consistently instead of `Failed` so we can consistently grep for error messages in logs.
```suggestion
						log.Error().Msgf("Error updating log level to %s: %s", logLevel, err)
```

File: cmd/osm-controller/osm-controller.go [hunk 5]
```diff
@@ -333,3 +337,44 @@ func getOSMControllerPod(kubeClient kubernetes.Interface) (*corev1.Pod, error) {
...
+					} else {
+						log.Info().Msgf("Global log level changed to: %s", logLevel)
+						currentLogLevel = logLevel
+					}
```
Comment (shashankram) [inline cmd/osm-controller/osm-controller.go:373]: nit: seems unnecessary to log this because we would otherwise see an error.

File: pkg/apis/config/v1alpha1/mesh_config.go [hunk 6]
```diff
@@ -91,6 +91,9 @@ type TrafficSpec struct {
 
 // ObservabilitySpec is the type to represent OSM's observability configurations.
 type ObservabilitySpec struct {
+	// OSMLogLevel defines the log level for OSM control plane logs
```
Comment (shashankram) [inline pkg/apis/config/v1alpha1/mesh_config.go:94]:
```suggestion
	// OSMLogLevel defines the log level for OSM control plane logs.
```

File: pkg/configurator/methods_test.go [hunk 7]
```diff
@@ -479,6 +482,25 @@ func TestCreateUpdateConfig(t *testing.T) {
 				assert.Equal(true, cfg.GetFeatureFlags().EnableAsyncProxyServiceMapping)
 			},
 		},
+		{
+			name: "OsmLogLevel",
```
Comment (shashankram) [inline pkg/configurator/methods_test.go:486]:
```suggestion
			name: "OSMLogLevel",
```

**Review by snehachhabria — APPROVED** [review]
Lets back port this to v0.9 as well ?

File: cmd/osm-controller/osm-controller.go (hunk 3 above)
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go:343]: thought of that. will do.

File: cmd/osm-controller/osm-controller.go (hunk 2 above)
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go:83]: ^ from commit msg
```
The initial log level flag has been kept as it still mandates boot log
level up to the point configurator and the log level handler are both
up, which then meshconfig value takes over.
```

File: cmd/osm-controller/osm-controller.go (hunk 4 above)
Comment (eduser25) [reply cmd/osm-controller/osm-controller.go:369]: good advice, changing

Comment (eduser25) [reply cmd/osm-controller/osm-controller.go:369]: will keep in mind also

**Review by shashankram — APPROVED** [review]

**Comment by eduser25** [comment]
@Mergifyio backport release-v0.9
